File: 15-3sum/3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        return nSum(nums, 3, 0, 0);
    }

    public List<List<Integer>> nSum (int[] nums, int n, int start, int target) {
        List<List<Integer>> res = new ArrayList<>();
        if (n < 2 || nums.length < n) return res;
        if (n == 2) {
            int lo = start, hi = nums.length - 1;
            while (lo < hi) {
                if (nums[lo] + nums[hi] == target) {
                    List<Integer> list = new ArrayList<>();
                    list.add(nums[lo]);
                    list.add(nums[hi]);
                    res.add(list);
                    lo++;
                    hi--;
                    while (lo < nums.length && nums[lo] == nums[lo - 1]) lo++;
                    while (hi >= 0 && nums[hi] == nums[hi + 1]) hi--;
                } else if (nums[lo] + nums[hi] < target) {
                    lo++;
                } else {
                    hi--;
                }
            }
        } else {
            for (int i = start; i < nums.length; i++) {
                if (i > 0 &&nums[i] == nums[i - 1]) continue;
                List<List<Integer>> subSum = nSum(nums, n - 1, i + 1, target - nums[i]);
                for (List<Integer> list : subSum) {
                    list.add(nums[i]);
                    res.add(list);
                }
            }
        }
        return res;
    }
}

File: 15-3sum/3sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        nums.sort()
        return self.nsum(nums, 3, 0, 0)
        
    def nsum(self, nums, n, start, target):
        sz = len(nums)
        res = []
        if n < 2 or sz < n:
            return []
        if n == 2:
            lo = start
            hi = sz - 1
            while lo < hi:
                if nums[lo] + nums[hi] == target:
                    res.append([nums[lo], nums[hi]])
                    lo += 1
                    while lo < sz and nums[lo] == nums[lo - 1]:
                        lo += 1
                    hi -= 1
                    while hi > lo and nums[hi] == nums[hi + 1]:
                        hi -= 1
                elif nums[lo] + nums[hi] < target:
                    lo += 1
                    while lo < sz and nums[lo] == nums[lo - 1]:
                        lo += 1
                else:
                    hi -= 1
                    while hi > lo and nums[hi] == nums[hi + 1]:
                        hi -= 1
            return res
        else:
            for i in range(start, sz):
                if i > start and nums[i] == nums[i - 1]:
                    continue
                sub_res = self.nsum(nums, n - 1, i + 1, target - nums[i])
                for arr in sub_res:
                    arr.append(nums[i])
                    res.append(arr)
            return res
            
